wiki/Writing-tests-with-good-behavioural-coverage.md

```diff
@@ -0,0 +1,147 @@
+
+For more details on testing methodologies specific to Oppia Android, please refer to the Oppia Testing Wiki.
+
+# Writing Effective Tests
+
+## 1. Understanding the Requirements
+
+Before writing tests thoroughly review documentation, user stories, or specifications for the functionality you're testing.
+**Example:**
+If you're testing a function that checks if a number is positive or negative, make sure it returns "Positive" for numbers greater than or equal to zero and "Negative" for numbers less than zero.
+
+```
+@Test
+fun testCheckSign_forPositiveInput_returnsPositive() {
+    assertEquals("Positive", checkSign(5))
+}
+
+@Test
+fun testCheckNumber_forNegativeInput_returnsNegative() {
+    assertEquals("Negative", checkSign(-3))
+}
+```
+
+# 2. Writing Clear and Descriptive Test Cases
+
+Each test case should:
+
+- Clearly describe the scenario and expected outcome.
+- Use descriptive names for your test methods.
+
+Naming Convention:
+```
+testAction_withOneCondition_withSecondCondition_hasExpectedOutcome
+```
+
+Example:
+```
+testCheckSign_forPositiveInput_returnsPositive()
+testCheckSign_forNegativeInput_returnsNegative()
+testCheckSign_forZeroInput_returnsNeitherPositiveNorNegative()
+```
+
+## 3. Covering Different Scenarios
+
+Ensure tests cover:
+
+- Positive Cases: Valid inputs that should pass.
+- Negative Cases: Invalid inputs or edge cases.
+- Boundary Cases: Inputs at the edge of acceptable ranges.
+
+```
+@Test
+fun testCheckSign_forZeroInput_returnsNeitherPositiveNorNegative() {
+  assertEquals("Neither Positive Nor Negative", checkSign(0)) // Boundary value
+}
+
+@Test
+fun testCheckSign_forPositiveBoundaryInput_returnsPositive() {
+  assertEquals("Positive", checkSign(1)) // Just above the boundary value
+}
+
+@Test
+fun testCheckSign_forNegativeBoundaryInput_returnsNegative() {
+  assertEquals("Negative", checkSign(-1)) // Just below the boundary value
+}
+```
+
+## 4. Exception Handling
+
+Exception Handling is a critical aspect of testing that ensures your code can handle and recover from error conditions gracefully.
+
+- Ensure Error Conditions Are Managed Gracefully:
+    - You need to verify that your application responds correctly when invalid or unexpected inputs are provided. For instance, if your function is designed to throw an exception when it receives null, you should test that it does so correctly.
+
+- Verify Correct Exception Types:
+    - Testing should confirm that the correct type of exception is thrown. This ensures that your error handling logic is specific and accurate.
+
+```
+@Test
+fun testCheckSign_forNullInput_throwsIllegalArgumentException() {
+    assertThrows<IllegalArgumentException> {
+        checkSign(null)
+    }
+}
+```
+
+- Check for Proper Exception Messages:
+    - It's important to ensure that exceptions include meaningful and informative messages that help diagnose issues. This can be tested by verifying the content of the exception message.
```

Comment:
Does it mean to check the actual error message with generic exceptions (eg: as mentioned IllegalStateException) to have a descriptive message. So instead of having a generic message like

```
throw IllegalStateException("Error") // Generic message
```

but to have a descriptive message

```
throw IllegalStateException("Cannot perform action: User account is inactive.") // Specific message
```

And does this cover this case? - https://github.com/oppia/oppia-android/blob/51393a49069bb14832eaebdba6178695f103437c/wiki/Writing-tests-with-good-behavioural-coverage.md#checking-for-proper-exception-messages or should I add something to actually differentiate what a generic message would (which is a don'ts) and specify the need for specific message (which is a dos). 

Will confirm my understanding on this and update as per feedback.

Reply:
I think your example touches on precisely what I was thinking, but it could maybe benefit from an additional note on why we care about verifying the message. While the existing reasoning is correct (that the message is clear), it's also especially necessary for generic exceptions to ensure the *correct* occurrence of that exception is being thrown. Consider a function that throws IllegalStateExceptions in two places (which is quite common since we use ``check`` in a bunch of places)--not verifying the error message could mean accidentally verifying the wrong exception is thrown which makes the test less useful as it's not actually verifying what it's meant to.

Reply:
(missed responding to this earlier) - Thanks for providing more context. I’ve added a section on the Importance of Specific Error Messages, based on the same reasoning: https://github.com/oppia/oppia-android/blob/a4c48d6dec31934b35fa4defafc113ad0c4f3661/wiki/Writing-tests-with-good-behavioural-coverage.md#importance-of-specific-error-messages